src/main/java/com/google/sps/servlets/DirectionsServlet.java

```diff
@@ -14,59 +14,63 @@
+import com.google.maps.DirectionsClientImpl;
 import java.io.IOException;
-import java.net.URISyntaxException;
 import java.util.Arrays;
 import java.util.List;
+
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
-import org.json.JSONException;
-import org.json.JSONObject;
 
+/**
+ * Serves key information from optimizing between addresses.
+ */
 @WebServlet("/directions")
 public class DirectionsServlet extends HttpServlet {
 
+  private final DirectionsClientFactory directionsClientFactory;
+
   /**
-   * Converts object to JSON and sends it to a HTTP Servlet Response.
-   *
-   * @param response Response to send object to.
-   * @param object Object to convert to JSON and send.
-   * @throws IOException
+   * Construct servlet with default DirectionsClient.
    */
-  public static void sendJson(HttpServletResponse response, Object object) throws IOException {
-    Gson gson = new Gson();
-    String json = gson.toJson(object);
-    response.setContentType("application/json");
-    response.getWriter().println(json);
+  public DirectionsServlet() {
+    directionsClientFactory = new DirectionsClientImpl.Factory();
+  }
+
+  /**
+   * Construct servlet with explicit implementation of DirectionsClient.
+   * @param factory A DirectionsClientFactory containing the implementation of DirectionsClientFactory.
+   */
+  public DirectionsServlet(DirectionsClientFactory factory) {
+    directionsClientFactory = factory;
   }
 
-  /** Get directions using the parameters stored in this class. */
+  /**
+   * Returns the most optimal order of travel between addresses.
+   * @param request HTTP request from the client.
+   * @param response HTTP response to the client.
+   * @throws IOException Indicates failed or interrupted I/O operations.
+   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String origin = request.getParameter("origin");
-    String destination = request.getParameter("destination");
-    String waypointsString = request.getParameter("waypoints");
-    String apiKey = request.getParameter("api-key");
-    List<String> waypoints;
-    if (waypointsString == null) {
-      waypoints = Arrays.asList();
-    } else {
-      waypoints = Arrays.asList(waypointsString.split(";"));
-    }
-    try {
-      JSONObject directions =
-          DirectionsUtility.getDirections(destination, origin, waypoints, apiKey);
-      sendJson(response, directions);
-    } catch (JSONException e) {
-      System.out.println(e);
-    } catch (URISyntaxException e) {
-      System.out.println(e);
-    } catch (IOException e) {
-      System.out.println(e + "; Remember to unrestrict IP address to use API Key.");
-    }
+    String apiKey = "AIzaSyBsHP0Wo698KQk2lkNlroMzSWHKyH9-05Y"; // TODO: Replace with get addresses from Tasks
```

Comment:
Probably wrong, but wasn't this api key supposed to be hidden?

Reply:
Absolutely, I have added a TODO to change this as I get to merging the ApiKeyProvider class. I have removed the API key from plain sight, but having this exposed here shouldn't be an issue though because it is heavily restricted.

Reply:
`RESOLVED`